`dapr run -f` docs (#3093)
dapr/docs: daprdocs/content/en/developing-applications/local-development/multi-app-dapr-run/multi-app-template.md

---
type: docs
title: "How to: Use the Multi-App Run template file"
linkTitle: "How to: Use the Multi-App Run template"
weight: 2000
description: Unpack the Multi-App Run template file and its properties
---

{{% alert title="Note" color="primary" %}}
 Multi-App Run is currently a preview feature only supported in Linux/MacOS. 
{{% /alert %}}

The Multi-App Run template file is a YAML file that you can use to run multiple applications at once. In this guide, you'll learn how to:
- Use the multi-app template 
- View started applications
- Stop the multi-app template 
- Stucture the multi-app template file

## Use the multi-app template

You can use the multi-app template file in one of the following two ways:

### Execute by providing a directory path

When you provide a directory path, the CLI will try to locate the Multi-App Run template file, named `dapr.yaml` by default in the directory. If the file is not found, the CLI will return an error.

Execute the following CLI command to read the Multi-App Run template file, named `dapr.yaml` by default:

```cmd

> // the template file needs to be called `dapr.yaml` by default

Reviewer: ```suggestion
# the template file needs to be called `dapr.yaml` by default if a directory path is given
```

dapr run -f <dir_path>
```

### Execute by providing a file path

If the Multi-App Run template file is named something other than `dapr.yaml`, then you can provide the relative or absolute file path to the command:

```cmd

> dapr run -f ./<your-preferred-file-name>.yaml

Reviewer: ```suggestion
dapr run -f ./path/to/<your-preferred-file-name>.yaml
```

```

## View the started applications

Once the multi-app template is running, you can view the started applications with the following command:

```cmd
dapr list
```

## Stop the multi-app template

Stop the multi-app run template anytime with either of the following commands:

```cmd

> dapr stop -f

Reviewer: ```suggestion
# the template file needs to be called `dapr.yaml` by default if a directory path is given

dapr stop -f <dir-path>
```

```
or:

```cmd

> dapr stop -f dapr.yaml

Reviewer: ```suggestion
dapr stop -f ./path/to/<your-preferred-file-name>.yaml
```

```

## Template file structure

The Multi-App Run template file can include the following properties. Below is an example template showing two applications that are configured with some of the properties. 

```yaml
version: 1
common: # optional section for variables shared across apps
  resourcesPath: ./app/components # any dapr resources to be shared across apps
  env:  # any environment variable shared across apps
    - DEBUG: true
apps:
  - appID: webapp

>     appDirPath: .dapr/webapp/ # required

Reviewer: ```suggestion
    appDirPath: .dapr/webapp/ # REQUIRED
```

    resourcesPath: .dapr/resources # (optional) can be default by convention
    configFilePath: .dapr/config.yaml # (optional) can be default by convention too, ignore if file is not found.
    appProtocol: HTTP
    appPort: 8080
    appHealthCheckPath: "/healthz" 
    command: ["python3" "app.py"]
  - appID: backend

>     appDirPath: .dapr/backend/

Reviewer: ```suggestion
    appDirPath: .dapr/backend/. # REQUIRED
```

| `logLevel`               | N        | The log verbosity. |  |
| `appMaxConcurrency`      | N        | The concurrency level of the application; default is unlimited |  |
| `placementHostAddress`   | N        |  |  |
| `appSSL`                 | N        | Enable https when Dapr invokes the application |  |
| `daprHTTPMaxRequestSize` | N        | Max size of the request body in MB. |  |
| `daprHTTPReadBufferSize` | N        | Max size of the HTTP read buffer in KB. This also limits the maximum size of HTTP headers. The default 4 KB |  |
| `enableAppHealthCheck`   | N        | Enable the app health check on the application | `true`, `false` |
| `appHealthCheckPath`     | N        | Path to the health check file | `/healthz` |
| `appHealthProbeInterval` | N        | Interval to probe for the health of the app in seconds
 |  |
| `appHealthProbeTimeout`  | N        | Timeout for app health probes in milliseconds |  |
| `appHealthThreshold`     | N        | Number of consecutive failures for the app to be considered unhealthy |  |
| `enableApiLogging`       | N        | Enable the logging of all API calls from application to Dapr |  |

> | `daprPath`               | N        | Dapr install path |  |

Reviewer: ```suggestion
| `runtimePath`               | N        | Dapr runtime install path |  |
```
Because of dapr/cli#1203

---
type: docs

> title: Multi-app template file

Reviewer 2: ```suggestion
title: Multi-app Run template file
```

linkTitle: Multi-app Run template
weight: 2000
description: Unpack the multi-app template file and its variables
---

{{% alert title="Note" color="primary" %}}
 Multi-app `dapr run -f` is currently a preview feature only supported in Linux/MacOS. 
{{% /alert %}}

> The multi-app template file is a single YAML configuration file that you can use to configure multiple applications alongside a Dapr sidecar. Execute the following command for Dapr to parse the multi-app template file, named `dapr.yaml` by default:

Reviewer 2: ```suggestion
The multi-app run template file is a YAML file that you can use to run multiple applications at once. Execute the following CLI command to read the multi-app template file, named `dapr.yaml` by default:
```

```cmd

> dapr run -f

Reviewer 2: ```suggestion
// the template file needs to be called `dapr.yaml` by default.
dapr run -f
```

```

> To name the multi-app template file something other than `dapr.yaml`, run:

Reviewer 2: ```suggestion
If the multi-app template file is named something other than `dapr.yaml`, run:
```

```cmd
dapr run -f ./<your-preferred-file-name>.yaml
```

> The multi-app template file can include any of the following parameters. 

Reviewer 2: ```suggestion
The multi-app template file can include the following properties. Below is an example template showing two applications that are configured with some of the properties.
```

```yaml
version: 1
common: # optional section for variables shared across apps
  resourcesPath: ./app/components # any dapr resources to be shared across apps
  env:  # any environment variable shared across apps
    - DEBUG: true
apps:
  - appID: webapp # required
    appDirPath: .dapr/webapp/ # required
    resourcesPath: .dapr/resources # (optional) can be default by convention
    configFilePath: .dapr/config.yaml # (optional) can be default by convention too, ignore if file is not found.
    appProtocol: HTTP
    appPort: 8080

>     appHealthCheckPath: "/healthz" # All _ converted to - for all properties defined under daprd section

Reviewer 2: Where did this comment come from 
"# All _ converted to - for all properties defined under daprd section"

Author: the example in one of the code PRs... should I remove the comment?

    command: ["python3" "app.py"]
  - appID: backend
    appDirPath: .dapr/backend/
    appProtocol: GRPC
    appPort: 3000
    unixDomainSocket: "/tmp/test-socket"
    env:
      - DEBUG: false
    command: ["./backend"]
```

## Parameters

| Parameter                | Required | Details | Example |
|--------------------------|:--------:|--------|---------|

> | `appID`                  | Y        | Your application's app ID | `webapp`, `backend` |

Reviewer 2: ```suggestion
| `appID`                  | Y        | Application's app ID | `webapp`, `backend` |
```

> | `appDirPath`             | Y        | Path to the your application | `./webapp/`, `./backend/` |

Reviewer 2: ```suggestion
| `appDirPath`             | Y        | Path to the your application code| `./webapp/`, `./backend/` |
```

| `resourcesPath`          | N        | Path to your Dapr resources. Can be default by convention; ignore if directory isn't found | `./app/components`, `./webapp/components` |
| `configFilePath`         | N        | Path to your application's configuration file | `./webapp/config.yaml` |

> | `appProtocol`            | N        | Application protocol | `HTTP`, `GRPC` |

Reviewer 2: ```suggestion
| `appProtocol`            | N        | The port your application is listening on | `HTTP`, `GRPC` |
```
Suggest that for many of the comments here you can take the same description as the table here https://docs.dapr.io/reference/cli/dapr-run/. I would also reference this table since they should be the same.

| `appPort`                | N        | Designated port for your application | `8080`, `3000` |
| `daprHTTPPort`           | N        | Dapr HTTP port |  |
| `daprGRPCPort`           | N        | Dapr GRPC port |  |
| `daprInternalGRPCPort`   | N        |  |  |

> | `metricsPort`            | N        |  |  |

Reviewer 2: ```suggestion
| `metricsPort`            | N        | The port that Dapr sends its metrics information to |  |
```

---
type: docs
title: MapR template file
linkTitle: MapR template
weight: 2000
description: Unpack the MapR template file and its properties
---

{{% alert title="Note" color="primary" %}}
 MapR (Multi-app Run using `dapr run -f`) is currently a preview feature only supported in Linux/MacOS. 
{{% /alert %}}

The MapR template file is a YAML file that you can use to run multiple applications at once. Execute the following CLI command to read the MapR template file, named `dapr.yaml` by default:

```cmd
// the template file needs to be called `dapr.yaml` by default

dapr run -f
```

If the MapR template file is named something other than `dapr.yaml`, run:

```cmd
dapr run -f ./<your-preferred-file-name>.yaml
```

The MapR template file can include the following properties. Below is an example template showing two applications that are configured with some of the properties. 

```yaml
version: 1
common: # optional section for variables shared across apps
  resourcesPath: ./app/components # any dapr resources to be shared across apps
  env:  # any environment variable shared across apps
    - DEBUG: true
apps:

>   - appID: webapp # required

Reviewer 3: correction: appID is not a required parameter. If it is not provided, it will be derived from `appDirPath`. In short, in the template file only the `appDirPath` is the only mandatory param.

---
type: docs
title: MapR template file
linkTitle: MapR template
weight: 2000
description: Unpack the MapR template file and its properties
---

{{% alert title="Note" color="primary" %}}
 MapR (Multi-app Run using `dapr run -f`) is currently a preview feature only supported in Linux/MacOS. 
{{% /alert %}}

> The MapR template file is a YAML file that you can use to run multiple applications at once. Execute the following CLI command to read the MapR template file, named `dapr.yaml` by default:

Reviewer 3: ```suggestion
## How to use
The MapR template file is a YAML file that you can use to run multiple applications at once. 
There are two ways to use the MapR template file:

#### 1. Execute by providing a directory path
When you provide a directory path, the CLI will try to locate the MapR template file, named `dapr.yaml` by default in the directory. If the file is not found, the CLI will return an error.
```

```cmd
// the template file needs to be called `dapr.yaml` by default

> dapr run -f

Reviewer 3: dapr run -f dir_path

> dapr run -f

Reviewer 3: ```suggestion
dapr run -f <dir_path>
```

```

> If the MapR template file is named something other than `dapr.yaml`, run:

Reviewer 3: ```suggestion
#### 2. Execute by providing a file path
If the MapR template file is named something other than `dapr.yaml` then you can provide the relative or absolute file path to the command:
```

```cmd
dapr run -f ./<your-preferred-file-name>.yaml
```

> The MapR template file can include the following properties. Below is an example template showing two applications that are configured with some of the properties. 

Reviewer 3: ```suggestion
## Template file structure
The MapR template file can include the following properties. Below is an example template showing two applications that are configured with some of the properties. 
```

```yaml
version: 1
common: # optional section for variables shared across apps
  resourcesPath: ./app/components # any dapr resources to be shared across apps
  env:  # any environment variable shared across apps
    - DEBUG: true
apps:
  - appID: webapp
    appDirPath: .dapr/webapp/ # required
    resourcesPath: .dapr/resources # (optional) can be default by convention
    configFilePath: .dapr/config.yaml # (optional) can be default by convention too, ignore if file is not found.
    appProtocol: HTTP
    appPort: 8080
    appHealthCheckPath: "/healthz" 
    command: ["python3" "app.py"]
  - appID: backend
    appDirPath: .dapr/backend/
    appProtocol: GRPC
    appPort: 3000
    unixDomainSocket: "/tmp/test-socket"
    env:
      - DEBUG: false
    command: ["./backend"]
```

> ## Template properties

Reviewer 3: ```suggestion
> Note: Below rules apply for all the paths present in the template file:
> - If the path is absolute, it is used as is.
> - All relative paths under comman section should be provided relative to the template file path.
> - `appDirPath` under apps section should be provided relative to the template file path.
> - All relative paths under app section should be provided relative to the appDirPath.

## Template properties
```
